### PR TITLE
Enable remote map download (#1723)

### LINKED ISSUE #1723: Add capability to download map via "custom protocol"
As discussed in https://github.com/triplea-game/triplea-game.github.io/issues/33 this would be a nice feature for the website.
We can configure install4j to write some values into the registry/create some files on linux based systems.
All we'd then need to do is to modify the main method so the maps are automatically downloaded when specific command line args are being passed.
@ssoloff @DanVanAtta How much work is needed to enable such a feature? Is it just a couple lines of code or do we need to change the download system before this feature can be implemented?

### PR DESCRIPTION
**Pull request by RoiEXLab**
This PR enables TripleA to download maps via the protocol implemented back in #1783
Basically this finally resolves #1723 
After this PR is merged, I will enable the download button on the TripleA website, to finalize this feature. (I will add a minimum required version notice though)

Known Issue:
Even if the Map is already downloaded, TripleA will still go ahead and download the map again.
@ssoloff If there is an easy solution to this problem (which also checks if the map is outdated), I will implement this fix as part of this PR (I couldn't find the code which checks if the map is already downloaded, but I think I was close), otherwhise I'llkindly ask you to submit a follow-up PR adressing this issue, which would notify the user that this map is already up-to-date

### CODE REVIEW

**Comment by ssoloff** [comment]
I've attached a [patch](https://github.com/triplea-game/triplea/files/1253900/pr-2274-ignore-installed-maps.patch.txt) that skips downloading maps that are already installed.

Since we were already notifying users if they attempted to download an unknown map, I augmented that dialog to also notify them if they attempted to download an installed map.  This dialog appears as follows if both an unknown and installed map were requested:

ignored-maps

If one type of map or the other was not requested, the corresponding text in the dialog will not be displayed.

**Comment by RoiEXLab** [comment]
@ssoloff Updated!

File: src/test/java/games/strategy/engine/framework/ArgParserTest.java [hunk 1]
```diff
@@ -59,11 +59,19 @@ public void malformedInputThrowsException() {
...
     ArgParser.handleCommandLineArgs(new String[] {TestData.propValue}, new String[] {GameRunner.TRIPLEA_GAME_PROPERTY});
     assertThat("if we pass only one arg, it is assumed to mean we are specifying the 'game property'",
         System.getProperty(GameRunner.TRIPLEA_GAME_PROPERTY), is(TestData.propValue));
   }
+  
+  @Test
+  public void singleUrlArgIsAssumedToBeMapDownloadProperty() {
+    final String testUrl = "triplea:" + TestData.propValue;
+    ArgParser.handleCommandLineArgs(new String[] {testUrl}, new String[] {GameRunner.TRIPLEA_MAP_DOWNLOAD_PROPERTY});
+    assertThat("if we pass only one arg, it is assumed to mean we are specifying the 'game property'",
```
Comment (ssoloff) [inline src/test/java/games/strategy/engine/framework/ArgParserTest.java]: Same failure message from `singleFileArgIsAssumedToBeGameProperty()`; it probably should be different.

Comment (RoiEXLab) [reply src/test/java/games/strategy/engine/framework/ArgParserTest.java]: The problem with copy pasting :D

**Comment by RoiEXLab** [comment]
@ssoloff I will adress this quickly ^^
In the meantime, please review my PRs on the website repo :P

**Comment by RoiEXLab** [comment]
@ssoloff Updated!

**Comment by prastle** [comment]
prob a dumb ? but how does this affect bots?
 @ssoloff and @RoiEXLab

**Comment by prastle** [comment]
can we automate them somehow?

**Comment by ssoloff** [comment]
@prastle 

> can we automate them somehow?

I don't think so.  This feature relies on the graphical game client.  The headless client used by the bots won't run the code to start the download that was added in this PR.

@RoiEXLab Please correct my statement about `GameRunner` vs. `HeadlessGameServer` if I'm wrong.

**Comment by ssoloff** [comment]
@prastle That's not to say that a similar feature couldn't be added to the headless client.  Just that it's not part of this PR.

**Comment by prastle** [comment]
np but yes they need to be automated
bots that is  
so we dont have to manually shut them down and re dl any map changes then relaunch them

**Comment by RoiEXLab** [comment]
@ssoloff @prastle  Unlikely we'll add such a feature (this PR) to the bots, as they are not installed via the installer.

Agreed we need to automate the bots somehow.
IMO there are 2 possible options: Either we have a daemon Thread in TripleA which checks regularily for map updates (inefficient, but would work even with our systemd setup, which automatically restarts them once they shut themselves down), or we add a CLI flag which makes headless bots check for map updates every time they are launched and download them then, a challenge would be to not have them all download the same maps at the same time (this would require us to start one at a time and then all the others after the maps have been downloaded).
In the latter case, we would just need to restart them regularily. @ssoloff Thoughts on how to potentially implement this?

**Comment by prastle** [comment]
Method 2 sounds best
Since we are often restarting them anyways ever few days

**Comment by prastle** [comment]
but present method for update means shutdown a server of bots reload them and then relaunch them.  This needs to be done to every bot every map update as @RoiEXLab  knows.  I am mainly typing this for @ssoloff in case he doesn't.  Thanks for the hard work guys!

**Comment by ssoloff** [comment]
> This needs to be done to every bot every map update

I'm assuming that was the solution submitted in #500.

If method 2 (download only when starting bot) works best for the admins, that's probably the easiest solution to implement, although not necessarily trivial.  All of the code that does the download heavy lifting is already written, it just needs a headless orchestration layer to take the place of `DownloadMapsWindow`.

I would suggest opening a new issue for this feature to discuss it further, as we're probably beyond the reasonable comment limit after a PR has been closed. :smile:

**Comment by RoiEXLab** [comment]
@ssoloff Yup, this is more or less the exact script, which is currently being used to download all maps.
It doesn't know which maps are actually out of date, so it just downloads them all again, which is really inefficient

**Comment by RoiEXLab** [comment]
@prastle In order to demonstrate what this PR does:
If you download the latest pre-release and install it, you can go to http://www.triplea-game.org/maps-list/all/ and download any map via your browser.

**Comment by prastle** [comment]
ok but another dumb ? but what is the difference between that and using in game downloads? or direct downloads?

**Comment by RoiEXLab** [comment]
There is none, this is just a nice little feature to show what is possible:
With more work it would be possible to have savegames load automatically over the web (by clicking on a link a savegame is automatically loaded), or to load custom maps (that are not in the official list).
I just wanted to show you, that this doesn't have anything to do with bots ^^

**Comment by RoiEXLab** [comment]
And well direct/manual download just gives you the zip file you'd have to put into your maps folder

**Comment by prastle** [comment]
Ok well ty and i luv ya but can we concentrate more on the broken stuff? ;)  but still great work if it can eventually be done from the cloud.

**Comment by DanVanAtta** [comment]
@prastle 
> np but yes they need to be automated
bots that is
so we dont have to manually shut them down and re dl any map changes then relaunch them

You can overwrite the maps in-place, the updated maps would take place the next time users change maps in the game hosting screen. It's not the best to do that, so often better to DL all maps to one folder, then move and overwrite that maps folder to the old one.

**Comment by RoiEXLab** [comment]
@DanVanAtta
@prastle meant that currently there's no trivial way to only update outdated maps.
Currently it's only possible to download all maps or each one manually.
It would be great if the bots would automatically update the maps on restarts.

**Comment by DanVanAtta** [comment]
Sure, though the existing process is easier than believed, the bots do not need to be stopped/started. The download_all script should be simplified a bit to have no extra steps and be made a one-liner. Today it's a two or three since you have to create a new folder and move it around. Regardless all that can be thrown into a cronjob and then we would not have to worry about it again. Ideally, we would make the 'download_all' script smarter to reduce DL amount, for the moment it does not matter so much

**Pull request merged**